Add magma gpu svd and matrix inverse

aam-at opened the pull request:
Adds magma gpu for computing svd and matrix inverse.

Benchmark results of magma library against lapack on 6 core cpu and Titan X. Results were taken from magma testing suite. Similar performance should be expected for theano.


<a id="orgfc86fbd"></a>

## SVD

<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-right" />

<col  class="org-right" />

<col  class="org-right" />
</colgroup>
<tbody>
<tr>
<td class="org-right">M</td>
<td class="org-right">CPU time (sec)</td>
<td class="org-right">GPU time (sec)</td>
</tr>


<tr>
<td class="org-right">1088</td>
<td class="org-right">0.1155</td>
<td class="org-right">0.2345</td>
</tr>


<tr>
<td class="org-right">2112</td>
<td class="org-right">0.5797</td>
<td class="org-right">0.5252</td>
</tr>


<tr>
<td class="org-right">3136</td>
<td class="org-right">2.5200</td>
<td class="org-right">1.1590</td>
</tr>


<tr>
<td class="org-right">4160</td>
<td class="org-right">6.1262</td>
<td class="org-right">2.1075</td>
</tr>


<tr>
<td class="org-right">5184</td>
<td class="org-right">15.5205</td>
<td class="org-right">4.0243</td>
</tr>
</tbody>
</table>


<a id="org50a95e3"></a>

## Matrix inverse

<table border="2" cellspacing="0" cellpadding="6" rules="groups" frame="hsides">


<colgroup>
<col  class="org-right" />

<col  class="org-left" />

<col  class="org-left" />
</colgroup>
<tbody>
<tr>
<td class="org-right">N</td>
<td class="org-left">CPU Gflop/s (sec)</td>
<td class="org-left">GPU Gflop/s (sec)</td>
</tr>


<tr>
<td class="org-right">1088</td>
<td class="org-left">78.01 (0.02)</td>
<td class="org-left">118.18 (0.01)</td>
</tr>


<tr>
<td class="org-right">2112</td>
<td class="org-left">132.95 (0.09)</td>
<td class="org-left">272.37 (0.05)</td>
</tr>


<tr>
<td class="org-right">3136</td>
<td class="org-left">135.28 (0.30)</td>
<td class="org-left">571.12 (0.07)</td>
</tr>


<tr>
<td class="org-right">4160</td>
<td class="org-left">139.91 (0.69)</td>
<td class="org-left">909.70 (0.11)</td>
</tr>


<tr>
<td class="org-right">5184</td>
<td class="org-left">141.60 (1.31)</td>
<td class="org-left">1358.48 (0.14)</td>
</tr>
</tbody>
</table>

nouiz commented on theano/gpuarray/linalg.py:
```diff
@@ -212,3 +217,129 @@ def perform(self, node, inputs, outputs):
+        if A.ndim != 2:
+            raise LinAlgError("Matrix rank error")
+        ctx_name = infer_context_name(A)
+        A = as_gpuarray_variable(A, ctx_name)
```
Move that line before the if. In case the user pass a list of list, ndim won't exist in that case.

aam-at replied:
Fixed.

nouiz commented on theano/gpuarray/linalg.py:
```diff
@@ -212,3 +217,129 @@ def perform(self, node, inputs, outputs):
+        if x.ndim != 2:
+            raise LinAlgError("Matrix rank error")
+        ctx_name = infer_context_name(x)
+        x = as_gpuarray_variable(x, ctx_name)
```
idem

nouiz commented on theano/gpuarray/magma_inv.c:
```diff
@@ -0,0 +1,101 @@
+#ifdef INPLACE
+  Py_XDECREF(out);
+  A_inv = A;
+  Py_INCREF(out);
```
out isn't defined. If the op get inplace, the compilation will crash.

aam-at replied:
Fixed.

nouiz commented on theano/gpuarray/tests/test_linalg.py:
```diff
@@ -112,3 +115,75 @@ def test_linalgerr_solve(self):
+
+
+class TestMagma(unittest.TestCase):
+    def test_gpu_matrix_inverse(self):
```
Make the tests skipped if magma isn't there.

Add a test for the inplace matrix_inverse

aam-at replied:
Fixed.

nouiz commented on theano/tensor/nlinalg.py:
```diff
@@ -625,9 +625,12 @@ def perform(self, node, inputs, outputs):
+                                           self.full_matrices,
+                                           self.compute_uv)
+        else:
+            u[0] = self._numop(x, self.full_matrices, self.compute_uv)
```
This change show that we need to test compute_uv. It is probably not tested, as the old code should crash.

Also, if we compute only 1 output, the make_node should have only 1 output.

aam-at replied:
I have numpy 1.12 and if `compute_uv = True` it returns only singular values. Which version of numpy does theano use for testing?

nouiz replied:
I don't know which version of numpy is installed, but I think your fix is good and don't depend of numpy version. What trouble me is that why tests didn't catch that crash? Do you have a Theano test that crash without that fix? I suppose not, so this would mean this case isn't tested.

aam-at replied:
@nouiz yeah, old tests do not check when `compute_uv = False`. How can I return only one output from `Apply` node without duplicating C code?

nouiz replied:
For the CPU op, as there is no C code, this isn't a problem.

For the GPU op, as it is a props, you can do it. I think this is an old questions that you found the answer.

Can you add a CPU test for that and fix the CPU make_node?

nouiz commented on theano/gpuarray/linalg.py:
```diff
@@ -212,3 +217,129 @@ def perform(self, node, inputs, outputs):
+    def __init__(self, inplace=False):
+        COp.__init__(self, ['magma_inv.c'],
+                     'APPLY_SPECIFIC(magma_inv)')
+        self.inplace = inplace
```
the inplace isn't working. Do you want to implement it? If so, you should add here:
if inplace:
    self.destry_map = {0:[0]}

And you should create an optimization to convert the non inplace version with the inplace version see the opt local_advincsub1_gpua_inplace in gpuarray/opt.py as an example.

The c code for inplace also need a update, I have another comment about that.

aam-at replied:
@nouiz I added inplace optimization.

nouiz commented:
Did you saw timing between cusolve svd and magma one?

aam-at commented:
@nouiz Thanks for the comments.

I used scikit-cuda demos to profile cusolver and magma. For `1000x1000` matrices, cusolver is about x2 times slower than magma. I didn't profile other matrix sizes.

Before starting this, I implemented cusolver svd with scikit-cuda in other branch [gpu_svd_op](https://github.com/aam-at/theano/tree/gpu_svd_op). `cusolver` interface is limited (only supports `full_matrices = True`; works only with wide matrices) and performance was unsatisfying. `magma` for matrix computations seems more promising and it is also used in other libraries for gpu matrix computations (e.g. pytorch).

nouiz commented:
thanks for the speed information. Can you confirm that when you timed cusolve, was this with a python interface as skcuda or it was compared with c code? Just the inteface itself could cause this difference.

It is good to know other framework use the same, it point that it is better for our cases. The sad part is that it make installation more complicated, so if using cusolve wasn't slower, for that reason it could be a better library to use. But that isn't the case.

aam-at commented:
@nouiz Both tested scikit-demos use python interface using ctypes.

Yeah, and unfortunately `magma` is not yet distributed via pip or ubuntu packages. I guess I also need to edit travis config to add installation and compilation for magma.

nouiz commented:
travis don't test the gpu, but our jenkins setup does. You won't have access to it, so it is to us to add it :)

aam-at commented on theano/configdefaults.py:
```diff
@@ -419,6 +419,23 @@ def f(suffix=suffix):
+AddConfigVar('magma.enabled',
+             " If True, use magma for matrix computation."
+             " If False, disable magma"
+             " If no_check, assume present and the version between header and library match (so less compilation at context init)",
```
Removed.

aam-at commented:
@nouiz I have addressed your comments. Is anything else necessary for this pull request?

nouiz commented on theano/configdefaults.py:
```diff
@@ -361,6 +361,22 @@ def default_dnn_library_path():
 
+AddConfigVar('magma.include_path',
+             "Location of the magma header",
+             StrParam(None),
```
I think None should be repalced by "". Otherwise I think it end up as 'None' as default value.

Also, they would need to be documented in doc/library/config.txt

nouiz commented on theano/gpuarray/linalg.py:
```diff
@@ -335,3 +340,146 @@ def perform(self, node, inputs, outputs):
+        compute_uv = int(self.compute_uv)
+        full_matrices = int(self.full_matrices)
+        return [('COMPUTE_UV', compute_uv),
+                ('FULL_MATRICES', full_matrices)]
```
I think you can easily add the infer_shape() method.

nouiz commented:
jenkins test this

notoraptor commented on theano/gpuarray/magma_inv.c:
```diff
@@ -0,0 +1,101 @@
+
+int APPLY_SPECIFIC(magma_inv)(PyGpuArrayObject *A, PyGpuArrayObject **_A_inv,
+                              PyGpuContextObject *c) {
+  PyGpuArrayObject *A_inv = *_A_inv;
```
I think you should always use `*_A_inv` instead of a temporar value `A_inv`. If a copy of `A_inv` is done (at line 47) but an error occurs after that and before setting `*_A_inv = A_inv`, then the memory allocated for the copy of `A_inv` will not be freed (as it is not checked at all at `fail` label) and will not be seen/handled by Theano (as `*_A_inv` will have not been updated).

aam-at commented:
@nouiz I added `infer_shape` to both cpu and gpu interface. I also updated `make_node` for both cpu and gpu versions.
@notoraptor I fixed that problem.

Jenkins currently skips `magma` tests.

notoraptor commented on theano/gpuarray/linalg.py:
```diff
@@ -335,3 +340,165 @@ def perform(self, node, inputs, outputs):
+        return shapes
+
+
+def gpu_matrix_inverse(a, inplace=False):
```
I think you should remove the argument `inplace`, as it is intended to be used internally by Theano for optimizations, not by users. Then you could just write:

`gpu_matrix_inverse = GpuMatrixInverse()`

aam-at replied:
Fixed.

notoraptor commented on theano/gpuarray/tests/test_linalg.py:
```diff
@@ -191,3 +199,100 @@ def test_invalid_input_fail_negative_definite(self):
+    def test_gpu_matrix_inverse(self):
+        A = theano.tensor.fmatrix("A")
+
+        fn = theano.function([A], gpu_matrix_inverse(A), mode=mode_with_gpu.including('magma'))
```
For all your tests, could you test if it works without `.including('magma')` (ie. just with `mode=mode_with_gpu`) ?

aam-at replied:
I removed that. My original idea was that the user can enable magma by enabling its optimizer. But approach with config `magma.enable` seems better.

notoraptor commented on theano/tensor/tests/test_nlinalg.py:
```diff
@@ -128,18 +129,47 @@ def test_qr_modes():
-    assert _allclose(n_v, t_v)
-    assert _allclose(n_t, t_t)
+    def test_svd(self):
+        rng = np.random.RandomState(utt.fetch_seed())
```
You could use `self.rng` instead, isn't it ?

aam-at replied:
Fixed.

notoraptor commented on theano/gpuarray/tests/test_linalg.py:
```diff
@@ -191,3 +199,97 @@ def test_invalid_input_fail_negative_definite(self):
+
+        fn = theano.function([A], gpu_matrix_inverse(A), mode=mode_with_gpu)
+        N = 1000
+        A_val = rand(N, N)
```
As your code only accepts float32 (`fmatrix` in Python, `GA_FLOAT` checking in C), always use float32 inputs, so that tests can pass even on 64 bits machines (where floatX is float64 by default). Here:
```python
A_val = rand(N, N).astype('float32')
```

aam-at replied:
Fixed.

notoraptor commented on theano/gpuarray/tests/test_linalg.py:
```diff
@@ -191,3 +199,97 @@ def test_invalid_input_fail_negative_definite(self):
+
+    def test_gpu_matrix_inverse_inplace(self):
+        N = 1000
+        A_val_gpu = gpuarray_shared_constructor(rand(N, N))
```
Same here (`rand(N, N).astype('float32')`).

notoraptor commented on theano/gpuarray/tests/test_linalg.py:
```diff
@@ -191,3 +199,97 @@ def test_invalid_input_fail_negative_definite(self):
+            np.dot(np.dot(U, S_m), VT), A, rtol=rtol, atol=atol)
+
+    def test_gpu_svd_wide(self):
+        A = rand(100, 50)
```
Same here.

notoraptor commented on theano/gpuarray/tests/test_linalg.py:
```diff
@@ -191,3 +199,97 @@ def test_invalid_input_fail_negative_definite(self):
+        self.assert_column_orthonormal(VT.T)
+
+    def test_gpu_svd_tall(self):
+        A = rand(50, 100)
```
Same here.

notoraptor commented on theano/gpuarray/tests/test_linalg.py:
```diff
@@ -191,3 +199,97 @@ def test_invalid_input_fail_negative_definite(self):
+        f_gpu = theano.function(
+            [A], gpu_svd(A, compute_uv=False), mode=mode_with_gpu)
+
+        A_val = rand(50, 100)
```
Same here.

notoraptor commented on theano/gpuarray/tests/test_linalg.py:
```diff
@@ -191,3 +199,97 @@ def test_invalid_input_fail_negative_definite(self):
+        A_val = rand(50, 100)
+        utt.assert_allclose(f_cpu(A_val), f_gpu(A_val))
+
+        A_val = rand(100, 50)
```
Same here.

notoraptor commented:
NB: There is a single conflict because of a PR recently merged that modifies an import statement in `gpuarray/opt.py`. Could you rebase your PR ?

aam-at commented:
@notoraptor I did rebase.

nouiz commented:
thanks!

botev commented on theano/gpuarray/magma_inv.c:
```diff
@@ -0,0 +1,99 @@
+    goto fail;
+  }
+
+  if (magma_imalloc_cpu(&piv, N)) {
```
Out of curiosity, why is the pivot array always allocated on the CPU?

aam-at replied:
`piv` is allocated on `cpu` in magma testing suite.
